nano/node/ledger_walker.cpp

```diff
@@ -0,0 +1,70 @@
+#include <nano/lib/blocks.hpp>
+#include <nano/lib/errors.hpp>
+#include <nano/node/ledger_walker.hpp>
+#include <nano/secure/blockstore.hpp>
+#include <nano/secure/ledger.hpp>
+
+#include <utility>
+
+nano::ledger_walker::ledger_walker (nano::ledger const & ledger_a) :
+	ledger{ ledger_a },
+	walked_blocks{},
+	blocks_to_walk{}
+{
+	if (ledger.store.init_error ())
+	{
+		throw nano::error{ "Ledger store initialization error" };
+	}
+}
+
+void nano::ledger_walker::walk_backward (nano::block_hash const & start_block_a, visitor_callback const & visitor_callback_a)
+{
+	const auto transaction = ledger.store.tx_begin_read ();
+
+	enqueue_block (start_block_a);
+	while (!blocks_to_walk.empty ())
+	{
+		const auto block = dequeue_block (transaction);
+		if (!visitor_callback_a (block))
+		{
+			continue;
+		}
+
+		if (!block->previous ().is_zero ())
+		{
+			enqueue_block (ledger.store.block_get (transaction, block->previous ()));
+		}
+
+		if (block->sideband ().details.is_receive)
+		{
+			enqueue_block (ledger.store.block_get (transaction, block->link ().as_block_hash ()));
+		}
+	}
+
+	walked_blocks.clear ();
```

Comment:
It doesn't seem right to clear walked_blocks and fully reset block_to_walk. I assume the swap is to free the memory. It seems like both should be swapped and put both of these in a function clear ().

It also needs a unit test to make sure multiple walks using a single ledger_walker instance works correctly.

Reply:
Additionally to the clear, a shrink_to_fit would do nicely, but I don't think that's part of the unordered_set API. Yeah, perhaps swapping both would be a good idea.